### PR TITLE
Fix TestConditionalProgressRequester and TestWaitUntilFreshAndListTimeout flakes

### LINKED ISSUE #119244: Failure cluster [7274a82c...] TestWaitUntilFreshAndListTimeout flakes
##### Error text:
```
Failed;
=== RUN   TestWaitUntilFreshAndListTimeout
W0711 14:40:52.951074   60838 logging.go:59] [core] [Channel #448 SubChannel #450] grpc: addrConn.createTransport failed to connect to {
  "Addr": "localhost:37881",
  "ServerName": "localhost",
  "Attributes": null,
  "BalancerAttributes": null,
  "Type": 0,
  "Metadata": null
}. Err: connection error: desc = "transport: Error while dialing: dial tcp [::1]:37881: connect: connection refused"
W0711 14:40:52.994944   60838 logging.go:59] [core] [Channel #460 SubChannel #463] grpc: addrConn.createTransport failed to connect to {
  "Addr": "localhost:45831",
  "ServerName": "localhost",
  "Attributes": null,
  "BalancerAttributes": null,
  "Type": 0,
  "Metadata": null
}. Err: connection error: desc = "transport: Error while dialing: dial tcp [::1]:45831: connect: connection refused"
W0711 14:40:53.499143   60838 logging.go:59] [core] [Channel #16 SubChannel #18] grpc: addrConn.createTransport failed to connect to {
  "Addr": "localhost:39985",
  "ServerName": "localhost",
  "Attributes": null,
  "BalancerAttributes": null,
  "Type": 0,
  "Metadata": null
}. Err: connection error: desc = "transport: Error while dialing: dial tcp [::1]:39985: connect: connection refused"
W0711 14:40:54.462863   60838 logging.go:59] [core] [Channel #388 SubChannel #390] grpc: addrConn.createTransport failed to connect to {
  "Addr": "localhost:41939",
  "ServerName": "localhost",
  "Attributes": null,
  "BalancerAttributes": null,
  "Type": 0,
  "
```
#### Recent failures:
[2023/7/12 05:24:16 ci-kubernetes-unit-eks-canary](https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-unit-eks-canary/1678877507660550144)
[2023/7/12 04:23:42 ci-kubernetes-unit](https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-unit/1678862155614523392)
[2023/7/12 03:22:52 ci-kubernetes-unit](https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-unit/1678846803849515008)
[2023/7/12 00:19:16 ci-kubernetes-unit-eks-canary](https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-unit-eks-canary/1678800750345457664)
[2023/7/11 22:18:13 ci-kubernetes-unit-eks-canary](https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-unit-eks-canary/1678770299090243584)

/kind flake

### PR DESCRIPTION
**Pull request by serathius**
/kind flake
```release-note
NONE
```
Fixes https://github.com/kubernetes/kubernetes/issues/119244
/assign @wojtek-t

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_progress_test.go [hunk 1]
```diff
@@ -48,17 +48,29 @@ func TestConditionalProgressRequester(t *testing.T) {
 	var requestsSent int32
 
 	logger.Info("No progress requests if no-one is waiting")
+	for !clock.HasWaiters() {
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_progress_test.go:51]: I'm not sure I'm following this - why the clock should have waiter? Can you add a comment explaining this?

**Comment by wojtek-t** [comment]
@serathius - did you run tests with stress as @aojea suggested elsewhere? If so, can you paste the output of that here?

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_progress_test.go (hunk 1 above)
Comment (serathius) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_progress_test.go:51]: Means Ticker was created. We want to wait for ticker to be created before we start progress the clock.

**Comment by serathius** [comment]
Compared flakiness with and without this change. With this PR:
* `TestWaitUntilFreshAndListTimeout` flakiness went from 21% to 0% (verified over 1k runs)
* `TestConditionalProgressRequester` flakiness went from 0.1% to 0% (verified over 6k runs)

Script used
```
apiserver $ go test ./pkg/storage/cacher/ -race -c
apiserver $ stress ./cacher.test --test.run 'TestWaitUntilFreshAndListTimeout'
```

**Comment by wojtek-t** [comment]
great - thanks!

/lgtm
/approve

**Pull request merged**